[WiP] Added HTTP Controller for Soundnode

wb9688 opened the pull request:
I've added an HTTP Controller as described in #798. It runs on port 8319. It's still a work in progress.

To do:
- [x] Add some CSS
- [x] Use WebSockets
- [x] Show current playlist
- [ ] Update playlist on change
- [x] Show track information
- [ ] Control volume
- [ ] Like button
- [ ] Play track of playlist on click
- [ ] Change progress on click
- [x] Make it an optional setting so user can switch the server on/off
- [ ] When user switch server on, display user IP + port (8319) so they can access the remote control
- [x] Test on a mobile device the remote control actually works

![Screenshot](http://i.imgur.com/10G1kjo.gif)

Pitros commented:
Fist of all it should be in separate files because there's going to be way more code - showing current track title, probably progress etc. 

Anyway are you going to update this pull request with more functions?

weblancaster commented:
Hey @wb9688 this is pretty cool but I wouldn't add this as a feature.

One of the reasons I built Soundnode is to have a Desktop experience without the need of a browser.

I appreciate the PR but for now this is not going to be accepted. thanks.

Pitros commented:
@weblancaster, this was supposed to be only a  possibility to control soundnode remotely from phone etc.

weblancaster commented:
@Pitros @jakejarrett please do not accept the PR I will work on with @wb9688  to make sure security and functionality are good.

Do keep contributing with whatever you guys want to.

weblancaster commented:
Yep @Pitros I took my comment back.. I'm just concern about security, UX and performance.

wb9688 commented:
@Pitros Yes, I'm going to add some more functions, but I don't know how to put it in a separate file as I don't know how Angular 1 works. Could someone help me with that? By the way I'm currently busy with adding some CSS...

Pitros commented:
@wb9688 @weblancaster uh I just realized that we would also should use websocket for connection between app and remote page. 
Performance impact won't be visible, but we're gaining a lot of possibilities. Socket.io package from npm is what we need, and all communication will be handled by sockets - no more redirects, only static page will be served.

wb9688 commented:
@Pitros That was also my idea, but Socket.io is really overkill. I'd suggest using `ws`. But as I said I don't know how to put it in a separate file as I don't know how Angular 1 works.

wb9688 commented:
I've added some CSS and now it looks like the following:

![Screenshot](https://cloud.githubusercontent.com/assets/9780201/16342758/0e9e4de6-3a34-11e6-9ee9-3ca6deff7e94.png)

weblancaster commented on app/public/js/player/playerCtrl.js:
```diff
@@ -358,4 +360,85 @@ app.controller('PlayerCtrl', function (
+                }
+                break;
+        }
+        res.end(`
```
do the same thing you did here but you will need to require the page e.g `var control =  require('control.html')` and then `res.end(control)`

weblancaster commented on app/public/js/player/playerCtrl.js:
```diff
@@ -358,4 +360,85 @@ app.controller('PlayerCtrl', function (
     });
 
 
+    var server = http.createServer(function (req, res) {
```
Move all this new code to a new folder and file e.g `app/public/js/control/control.js` `app/public/view/control/control.html`

weblancaster commented:
Base the style on what we have on the desktop app.. you can just reuse them and refactor

wb9688 commented:
@weblancaster And should I do something in index.html? And how should I create the Angular controller for accessing the `playerService`?

weblancaster commented:
You can follow the same way it's done for other features. You will need to add the files to the index.html on the respective place

https://github.com/Soundnode/soundnode-app/blob/master/app/index.html#L203

wb9688 commented:
OK, I made a `remote-control.js` with the following content:

```js
'use strict';

var fs = require('fs');
var http = require('http');
var remoteControl = fs.readFileSync('./views/remote-control/remote-control.html');

app.controller('RemoteCtrl', function (
    $scope,
    $rootScope,
    playerService,
    mprisService,
    queueService,
    hotkeys,
    $state,
    $log,
    $timeout,
    SCapiService,
    notificationFactory
) {
    var server = http.createServer(function (req, res) {
        switch (req.url) {
            case '/playpause':
                if ($rootScope.isSongPlaying) {
                    playerService.pauseSong();
                } else {
                    playerService.playSong();
                }
                break;
            case '/prev':
                if ($rootScope.isSongPlaying) {
                    playerService.playPrevSong();
                }
                break;
            case '/next':
                if ($rootScope.isSongPlaying) {
                    playerService.playNextSong();
                }
                break;
        }
        res.end(remoteControl);
    });
    server.listen(8319);
});
```

And added `<script src="public/js/remote-control/remote-control.js"></script>` under `<script src="public/js/news/newsCtrl.js"></script>`. I also removed the code from `playerCtrl.js`.

But I get an `Unable to connect` in Firefox, so it doesn't work. Do I need to register it somewhere else? Sorry, I'm a total Angular 1 noob...

wb9688 commented:
@weblancaster There's no reliable way to detect the user's IP address, because in Node.js you get a list of your network interface with `os.networkInterfaces()` and on my laptop I get [this output](https://gist.github.com/wb9688/524bb0c7f74569e279308493cd791bad). And my IP address is `192.168.1.127` and `192.168.56.1` is a virtual IP address of VirtualBox...

wb9688 commented:
Now it looks like the following:

![Screenshot](https://cloud.githubusercontent.com/assets/9780201/16356179/9b06e0c2-3acf-11e6-8c9a-0bc3ea7152d2.png)

weblancaster commented:
Great progress @wb9688 

For the IP you can use https://github.com/indutny/node-ip

wb9688 commented:
@weblancaster I had also found that module, but it returns the wrong IP
address on my laptop…

weblancaster commented:
So we got figure out a way that will work for all platforms (Mac/Linux/Win)

weblancaster commented:
closing this since no more work is being done here.

wb9688 commented:
@weblancaster: I've been busy, but now I've some free time again. I've started with adding a like button to the branch. So, could you please re-open this PR?

weblancaster commented:
I'm migrating from nwjs to electronjs and might affect this work @wb9688

wb9688 commented:
@weblancaster: If you're going to migrate to Electron, I just need to place
the code in other places…
